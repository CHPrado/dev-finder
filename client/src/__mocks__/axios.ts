const axios = {
  get: jest.fn(() => Promise.resolve({})),
  post: jest.fn(() => Promise.resolve({})),
  create: () => axios,
  defaults: {
    adapter: {},
  },
};

export default axios;
