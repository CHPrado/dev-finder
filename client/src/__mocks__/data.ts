import { createMemoryHistory, createLocation } from "history";

import { DevProps, DevRouteProps } from "../interfaces";

const devData: DevProps = {
  username: "CHPrado",
  name: "Christian Prado",
  bio: "My bio",
  avatar_url: "avatar_url",
  location: "Lins - SP",
  followers: 1,
  following: 2,
  public_repos: 3,
  stars: 4,
  watchers: 5,
  forks: 6,
  languages: [
    {
      name: "JavaScript",
      count: 1,
      percent: "49.9",
    },
    {
      name: "Ruby",
      count: 1,
      percent: "51.1",
    },
  ],
  notes: "Notes about me",
};

const devRouteData: DevRouteProps = {
  history: createMemoryHistory(),
  location: createLocation({ state: { dev: devData } }),
  match: {
    isExact: false,
    path: "/dev",
    url: "/dev",
    params: {},
  },
};

export { devData, devRouteData };
