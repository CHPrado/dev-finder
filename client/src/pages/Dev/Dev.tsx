import React, { FC, useState } from "react";
import { Link } from "react-router-dom";

import {
  Box,
  Card,
  CircularProgress,
  IconButton,
  Tooltip,
  Typography,
} from "@material-ui/core";
import { CachedOutlined } from "@material-ui/icons";
import { LocationIcon } from "@primer/octicons-react";

import { Header, Snackbar } from "../../components";
import { SnackbarProps, DevRouteProps } from "../../interfaces";
import { githubApi } from "../../services";
import { Details } from "./Details";
import { Languages } from "./Languages";
import { Notes } from "./Notes";
import useStyles from "./styles";

const Dev: FC<DevRouteProps> = (props) => {
  const classes = useStyles();
  const [dev, setDev] = useState(props.location.state.dev);
  const [isLoading, setIsLoading] = useState(false);
  const [snackbar, setSnackbar] = useState<SnackbarProps>({});

  const handleRefresh = () => {
    setIsLoading(true);
    githubApi
      .getDev(dev.username)
      .then((resp) => {
        setDev(resp);
      })
      .catch(() => {
        setSnackbar({
          open: true,
          message:
            "There was an error trying to refresh the data. Please, try again.",
          type: "error",
          onClose: () => setSnackbar({ open: false }),
        });
      })
      .finally(() => setIsLoading(false));
  };

  return (
    <Box className={classes.root}>
      <Header />

      {isLoading ? (
        <Box className={classes.loadingBox}>
          <CircularProgress size="100px" />
        </Box>
      ) : (
        <Box className={classes.container}>
          <Card variant="outlined" className={classes.card}>
            <Box className={classes.titleBox}>
              <Link
                className={classes.link}
                to={(location) => ({
                  ...location,
                  pathname: `http://www.github.com/${dev.username}`,
                })}
                target="_blank"
              >
                <Typography className={classes.name}>
                  {`${dev.username}'s profile`}
                </Typography>
              </Link>

              <Box display="flex">
                <Tooltip title="Refresh">
                  <IconButton onClick={handleRefresh}>
                    <CachedOutlined />
                  </IconButton>
                </Tooltip>
              </Box>
            </Box>

            <Box className={classes.avatarBox}>
              <img
                src={dev.avatar_url}
                alt="avatar"
                className={classes.avatar}
              />

              <Box className={classes.infoBox}>
                <Typography className={classes.info}>{dev.name}</Typography>
                {dev.location && (
                  <Typography className={classes.info}>
                    <LocationIcon size={18} className={classes.icon} />
                    {dev.location}
                  </Typography>
                )}
                <Typography color="textSecondary" align="justify">
                  {dev.bio}
                </Typography>
              </Box>
            </Box>

            <Box className={classes.detailsBox}>
              <Details dev={dev} />
              <Languages languages={dev.languages} />
            </Box>

            <Notes
              dev={dev}
              setIsLoading={setIsLoading}
              setSnackbar={setSnackbar}
              setDev={setDev}
            />
          </Card>
        </Box>
      )}

      <Snackbar {...snackbar} />
    </Box>
  );
};

export default Dev;
