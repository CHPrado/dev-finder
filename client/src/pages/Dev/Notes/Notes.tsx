import React, {
  ChangeEvent,
  Dispatch,
  FC,
  SetStateAction,
  useState,
} from "react";

import {
  Box,
  Button,
  IconButton,
  TextField,
  Tooltip,
  Typography,
} from "@material-ui/core";
import { EditOutlined } from "@material-ui/icons";

import { DevProps, SnackbarProps } from "../../../interfaces";
import { devsApi } from "../../../services";
import useStyles from "./styles";

interface Props {
  dev: DevProps;
  setIsLoading: Dispatch<SetStateAction<boolean>>;
  setSnackbar: Dispatch<SetStateAction<SnackbarProps>>;
  setDev: Dispatch<SetStateAction<DevProps>>;
}

const Notes: FC<Props> = ({ dev, setIsLoading, setSnackbar, setDev }) => {
  const classes = useStyles();
  const [edit, setEdit] = useState(false);
  const [notes, setNotes] = useState(dev.notes);

  const handleClickEdit = () => {
    setEdit(true);
  };

  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    setNotes(event.target.value);
  };

  const handleCancel = () => {
    setNotes(dev.notes || "");
    setEdit(false);
  };

  const handleSave = () => {
    setIsLoading(true);
    devsApi
      .save(dev.username, notes)
      .then((resp) => {
        setDev({ ...dev, notes });
        setSnackbar({
          open: true,
          message: resp?.data?.message || "Notes saved!",
          type: "success",
          onClose: () => setSnackbar({ open: false }),
        });
      })
      .catch(() => {
        setSnackbar({
          open: true,
          message: "There was an error to save the notes. Please, try again.",
          type: "error",
          onClose: () => setSnackbar({ open: false }),
        });
      })
      .finally(() => {
        setIsLoading(false);
        setEdit(false);
      });
  };

  return (
    <Box width="100%">
      <Box display="flex" justifyContent="space-between" alignItems="center">
        <Typography color="textSecondary" variant="h6">
          Notes
        </Typography>
        {!edit && (
          <Tooltip title="Edit note">
            <IconButton onClick={handleClickEdit}>
              <EditOutlined />
            </IconButton>
          </Tooltip>
        )}
      </Box>
      <TextField
        placeholder={`Add notes about ${dev.name || dev.login}`}
        value={notes}
        onChange={handleChange}
        className={classes.notes}
        variant="outlined"
        multiline
        disabled={!edit}
        rows={5}
        fullWidth
      />
      {edit && (
        <Box className={classes.buttonsBox}>
          <Button
            className={classes.button}
            variant="contained"
            color="primary"
            onClick={handleSave}
          >
            SAVE
          </Button>
          <Button
            className={classes.button}
            variant="outlined"
            color="primary"
            onClick={handleCancel}
          >
            CANCEL
          </Button>
        </Box>
      )}
    </Box>
  );
};

export default Notes;
