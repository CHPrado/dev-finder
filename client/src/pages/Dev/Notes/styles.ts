import { Theme, makeStyles } from "@material-ui/core/styles";

export default makeStyles((theme: Theme) => ({
  notes: {
    margin: theme.spacing(5, 0),
  },
  buttonsBox: {
    display: "flex",
    [theme.breakpoints.down("xs")]: {
      display: "flex",
      alignItems: " center",
      justifyContent: "center",
      flexDirection: "column",
    },
  },
  button: {
    marginRight: theme.spacing(2),
    width: "150px",
    [theme.breakpoints.down("xs")]: {
      flexDirection: "column",
      marginTop: theme.spacing(2),
      width: "100%",
    },
  },
}));
