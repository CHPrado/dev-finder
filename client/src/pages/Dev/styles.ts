import { Theme, makeStyles } from "@material-ui/core/styles";

const flexCenter = {
  display: "flex",
  alignItems: "center",
  justifyContent: "center",
};

export default makeStyles((theme: Theme) => ({
  root: {
    display: "flex",
    flexDirection: "column",
    height: "100%",
  },
  loadingBox: {
    height: "100%",
    ...flexCenter,
  },
  container: {
    ...flexCenter,
    position: "absolute",
    top: "100px",
    width: "100%",
  },
  card: {
    ...flexCenter,
    flexDirection: "column",
    margin: theme.spacing(5),
    padding: theme.spacing(5),
    maxWidth: "500px",
    [theme.breakpoints.down("xs")]: {
      width: "100%",
    },
  },
  titleBox: {
    width: "100%",
    display: "flex",
    alignItems: "center",
    justifyContent: "space-between",
    marginBottom: theme.spacing(4),
  },
  name: {
    color: theme.palette.primary.dark,
    fontSize: theme.typography.h6.fontSize,
    fontWeight: theme.typography.fontWeightBold,
    whiteSpace: "nowrap",
  },
  avatarBox: {
    display: "flex",
    alignItems: "flex-start",
    justifyContent: "center",
    [theme.breakpoints.down("xs")]: {
      flexDirection: "column",
      alignItems: "center",
    },
  },
  avatar: {
    borderRadius: "50%",
    height: "150px",
    marginRight: theme.spacing(4),
    [theme.breakpoints.down("xs")]: {
      height: "200px",
      marginBottom: theme.spacing(4),
    },
  },
  infoBox: {
    marginBottom: theme.spacing(4),
    display: "flex",
    flexDirection: "column",
    width: "100%",
  },
  info: {
    color: theme.palette.text.secondary,
    fontSize: theme.typography.h6.fontSize,
    fontWeight: theme.typography.fontWeightBold,
    display: "flex",
    alignItems: "center",
  },
  detailsBox: {
    width: "100%",
    display: "flex",
    justifyContent: "center",
    margin: theme.spacing(4),
    [theme.breakpoints.down("xs")]: {
      flexDirection: "column",
    },
  },
  icon: {
    marginRight: theme.spacing(2),
  },
  link: {
    textDecoration: "none",
    "& :hover": {
      color: theme.palette.primary.main,
      textDecoration: "underline",
    },
  },
}));
