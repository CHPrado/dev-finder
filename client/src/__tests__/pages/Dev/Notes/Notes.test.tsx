import React from "react";

import { Button, IconButton, TextField } from "@material-ui/core";
import { act } from "@testing-library/react";
import axios from "axios";
import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";

import { devData } from "../../../../__mocks__/data";
import { Notes } from "../../../../pages/Dev/Notes";

jest.mock("axios");
const setDev = jest.fn();
const setIsLoading = jest.fn();
const setSnackbar = jest.fn();

const wrapper = mount(
  <Notes
    dev={devData}
    setDev={setDev}
    setIsLoading={setIsLoading}
    setSnackbar={setSnackbar}
  />
);

describe("rendering component", () => {
  it("renders Notes component without crashing", () => {
    const wrapper = shallow(
      <Notes
        dev={devData}
        setDev={setDev}
        setIsLoading={setIsLoading}
        setSnackbar={setSnackbar}
      />
    );
    const notes = wrapper.find(TextField).at(0).props().value;

    expect(notes).toEqual("Notes about me");
  });
});

describe("passing props", () => {
  it("dev props data equals to dev data", () => {
    expect(wrapper.props().dev).toEqual(devData);
    expect(wrapper.props().setDev).toEqual(setDev);
    expect(wrapper.props().setIsLoading).toEqual(setIsLoading);
    expect(wrapper.props().setSnackbar).toEqual(setSnackbar);
  });
});

describe("simulating events", () => {
  describe("simulating edit button click", () => {
    let notes = wrapper.find(TextField).at(0);

    it("checks notes TextField prop before click", () => {
      expect(notes.props().disabled).toEqual(true);
    });

    describe("simulating edit button click", () => {
      const editButton = wrapper.find(IconButton).at(0);
      editButton.simulate("click");
      wrapper.update();

      it("checks notes TextField prop after click", () => {
        notes = wrapper.find(TextField).at(0);
        expect(notes.props().disabled).toEqual(false);
      });
    });

    describe("simulating cancel button click", () => {
      const cancelButton = wrapper.find(Button).at(1);
      expect(cancelButton.render().text()).toEqual("CANCEL");

      it("simulates cancel button click", () => {
        cancelButton.simulate("click");
        wrapper.update();
        notes = wrapper.find(TextField).at(0);

        expect(notes.props().disabled).toEqual(true);
        expect(setDev).toBeCalledTimes(0);
        expect(setIsLoading).toBeCalledTimes(0);
        expect(setSnackbar).toBeCalledTimes(0);
      });
    });

    describe("simulating save button click", () => {
      const saveButton = wrapper.find(Button).at(0);
      expect(saveButton.render().text()).toEqual("SAVE");

      it("simulates save button click", async () => {
        const getSpy = jest.spyOn(axios, "post");

        await act(async () => {
          saveButton.simulate("click");
        });

        wrapper.update();
        notes = wrapper.find(TextField).at(0);

        expect(notes.props().disabled).toEqual(true);
        expect(setDev).toBeCalled();
        expect(setIsLoading).toBeCalled();
        expect(setSnackbar).toBeCalled();
        expect(getSpy).toBeCalled();
      });
    });
  });
});

describe("snapshots", () => {
  it("Notes snapshots", () => {
    const tree = shallow(
      <Notes
        dev={devData}
        setDev={setDev}
        setIsLoading={setIsLoading}
        setSnackbar={setSnackbar}
      />
    );
    expect(toJson(tree)).toMatchSnapshot();
  });
});
