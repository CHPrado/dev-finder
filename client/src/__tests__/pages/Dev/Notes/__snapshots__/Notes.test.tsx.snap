// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots Notes snapshots 1`] = `
<Styled(MuiBox)
  width="100%"
>
  <Styled(MuiBox)
    alignItems="center"
    display="flex"
    justifyContent="space-between"
  >
    <WithStyles(ForwardRef(Typography))
      color="textSecondary"
      variant="h6"
    >
      Notes
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Tooltip))
      title="Edit note"
    >
      <WithStyles(ForwardRef(IconButton))
        onClick={[Function]}
      >
        <Memo([object Object]) />
      </WithStyles(ForwardRef(IconButton))>
    </WithStyles(ForwardRef(Tooltip))>
  </Styled(MuiBox)>
  <WithStyles(ForwardRef(TextField))
    className="makeStyles-notes-1"
    disabled={true}
    fullWidth={true}
    multiline={true}
    onChange={[Function]}
    placeholder="Add notes about Christian Prado"
    rows={5}
    value="Notes about me"
    variant="outlined"
  />
</Styled(MuiBox)>
`;
